### PR TITLE
macOS port

### PR DESCRIPTION
**Pull request by apangin**

### CODE REVIEW

File: src/symbols_macos.cpp [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+
+#include <mach-o/loader.h>
+#include <cstring>
+#include <mach-o/nlist.h>
+#include "symbols.h"
+#include <sys/mman.h>
+#include <mach-o/dyld_images.h>
+#include <fcntl.h>
+
+// Workaround for newer macosx versions: since 10.12.x _dyld_get_all_image_infos exists, but is not exported
```
Comment (retronym) [inline src/symbols_macos.cpp]: I'm getting a linker error in High Sierra.

```
⚡ xcode-select -p; uname -a ; clang --version; make clean all
/Applications/Xcode.app/Contents/Developer
Darwin iMac2016.local 17.0.0 Darwin Kernel Version 17.0.0: Thu Aug 24 21:48:19 PDT 2017; root:xnu-4570.1.46~2/RELEASE_X86_64 x86_64
Apple LLVM version 9.0.0 (clang-900.0.38)
Target: x86_64-apple-darwin17.0.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
rm -rf build
mkdir -p build
g++ -O2 -D_XOPEN_SOURCE -D_DARWIN_C_SOURCE -I/Library/Java/JavaVirtualMachines/jdk1.8.0_144.jdk/Contents/Home/include -I/Library/Java/JavaVirtualMachines/jdk1.8.0_144.jdk/Contents/Home/include/darwin -fPIC -shared -o build/libasyncProfiler.so src/*.cpp -ldl -lpthread
Undefined symbols for architecture x86_64:
  "__dyld_get_all_image_infos", referenced from:
      Symbols::parseMaps(NativeCodeCache**, int) in symbols_macos-0787ab.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [build/libasyncProfiler.so] Error 1
```

More verbosely:

```
 "/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ld" -demangle -lto_library /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/libLTO.dylib -dynamic -dylib -arch x86_64 -macosx_version_min 10.13.0 -o build/libasyncProfiler.so /var/folders/b7/xcc2k0ln6ldcv247ffpy2d1w0000gp/T/allocTracer-ddbd60.o /var/folders/b7/xcc2k0ln6ldcv247ffpy2d1w0000gp/T/arguments-a50ec6.o /var/folders/b7/xcc2k0ln6ldcv247ffpy2d1w0000gp/T/codeCache-b3d94d.o /var/folders/b7/xcc2k0ln6ldcv247ffpy2d1w0000gp/T/frameName-40c299.o /var/folders/b7/xcc2k0ln6ldcv247ffpy2d1w0000gp/T/javaApi-49f7b2.o /var/folders/b7/xcc2k0ln6ldcv247ffpy2d1w0000gp/T/lockTracer-0848b7.o /var/folders/b7/xcc2k0ln6ldcv247ffpy2d1w0000gp/T/perfEvents_linux-5444f2.o /var/folders/b7/xcc2k0ln6ldcv247ffpy2d1w0000gp/T/perfEvents_macos-45665a.o /var/folders/b7/xcc2k0ln6ldcv247ffpy2d1w0000gp/T/profiler-4e1680.o /var/folders/b7/xcc2k0ln6ldcv247ffpy2d1w0000gp/T/stackFrame_aarch64-07472e.o /var/folders/b7/xcc2k0ln6ldcv247ffpy2d1w0000gp/T/stackFrame_arm-6d17de.o /var/folders/b7/xcc2k0ln6ldcv247ffpy2d1w0000gp/T/stackFrame_i386-909c53.o /var/folders/b7/xcc2k0ln6ldcv247ffpy2d1w0000gp/T/stackFrame_x64-5382a2.o /var/folders/b7/xcc2k0ln6ldcv247ffpy2d1w0000gp/T/symbols_linux-b646fd.o /var/folders/b7/xcc2k0ln6ldcv247ffpy2d1w0000gp/T/symbols_macos-b23600.o /var/folders/b7/xcc2k0ln6ldcv247ffpy2d1w0000gp/T/vmEntry-e4f62e.o /var/folders/b7/xcc2k0ln6ldcv247ffpy2d1w0000gp/T/vmStructs-fa938e.o -ldl -lpthread -lc++ -lSystem /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/9.0.0/lib/darwin/libclang_rt.osx.a
Undefined symbols for architecture x86_64:
  "__dyld_get_all_image_infos", referenced from:
      Symbols::parseMaps(NativeCodeCache**, int) in symbols_macos-b23600.o
ld: symbol(s) not found for architecture x86_64
```

Is that expected or is it something I can remedy?

Comment (retronym) [reply src/symbols_macos.cpp]: Related discussion about the API changes: https://bugs.chromium.org/p/crashpad/issues/detail?id=187

Comment (apangin) [reply src/symbols_macos.cpp]: I've commited the fix. Please check if works for you.

Comment (retronym) [reply src/symbols_macos.cpp]: Thanks! I was able to build and run the `-e cpu` profile.

```
⚡ ./profiler.sh collect -e cpu -b 5000000 -d 30 -o summary $(jps | grep ForkedMain | awk -d' ' -e '{print $1}')
Started [cpu] profiling
--- Execution profile ---
Total samples:         14590
Non-Java:              252 (1.73%)
Unknown (native):      2492 (17.08%)
Unknown (Java):        144 (0.99%)
Not walkable (Java):   1 (0.01%)

Frame buffer usage:    17.6754%
```

I take it that the Mac profiler currently only understands Java frames? My app ends up calling native code for IO, Zip reading, and hashCodes a lot, which probably accounts for the unknown native frames above. 

If a sample occurs in a native frame, will that sample be attributed to the Java portion of the stack that calls the native code?

For reference, here's the corresponding `-Xprof`

```
[info] Flat profile of 209.87 secs (17030 total ticks): scala.tools.nsc.HotScalacBenchmark.compile-jmh-worker-1
[info]
[info]   Interpreted + native   Method
[info]   0.1%    15  +     0    scala.reflect.internal.tpe.TypeMaps$TypeMap.mapOver
[info]   0.1%    11  +     0    java.lang.ClassLoader.defineClass1
[info]   0.1%     9  +     0    scala.reflect.internal.Trees.itransform
[info]   0.0%     8  +     0    scala.tools.nsc.typechecker.Contexts$Context.make
[info]   0.0%     6  +     0    scala.tools.nsc.transform.Erasure$Eraser.typed1
[info]   0.0%     5  +     0    scala.tools.nsc.typechecker.RefChecks$RefCheckTransformer.transform
[info]   0.0%     5  +     0    scala.tools.nsc.transform.UnCurry$UnCurryTransformer.$anonfun$postTransform$1
[info]   0.0%     2  +     3    sun.nio.fs.UnixNativeDispatcher.open0
[info]   0.0%     5  +     0    scala.tools.nsc.typechecker.Macros.macroExpandWithRuntime
[info]   0.0%     4  +     0    scala.tools.nsc.PhaseAssembly.computePhaseAssembly
[info]   0.0%     4  +     0    scala.tools.nsc.Global$Run.compileUnitsInternal
[info]   0.0%     4  +     0    scala.tools.nsc.typechecker.Typers$Typer.typedInAnyMode$1
[info]   0.0%     4  +     0    scala.reflect.internal.StdNames$TermNames.<init>
[info]   0.0%     4  +     0    scala.tools.nsc.typechecker.Unapplies.caseModuleUnapplyMeth
[info]   0.0%     3  +     0    scala.tools.nsc.typechecker.SyntheticMethods.$anonfun$addSyntheticMethods$19
[info]   0.0%     3  +     0    scala.tools.nsc.typechecker.SyntheticMethods.$anonfun$addSyntheticMethods$8
[info]   0.0%     3  +     0    scala.tools.nsc.typechecker.TypeDiagnostics$TyperDiagnostics$checkDead$.<init>
[info]   0.0%     3  +     0    scala.reflect.internal.BaseTypeSeqs.compoundBaseTypeSeq
[info]   0.0%     3  +     0    scala.reflect.internal.HasFlags.isMacro
[info]   0.0%     3  +     0    scala.tools.nsc.settings.ScalaSettings.$init$
[info]   0.0%     3  +     0    scala.tools.nsc.typechecker.Implicits$ImplicitSearch.companionImplicitMap
[info]   0.0%     3  +     0    scala.tools.nsc.ast.parser.Parsers$Parser.convertToParams
[info]   0.0%     0  +     3    sun.misc.Unsafe.defineAnonymousClass
[info]   0.0%     3  +     0    scala.reflect.internal.Definitions$DefinitionsClass.BoxedByteClass$lzycompute
[info]   0.0%     3  +     0    scala.reflect.internal.SymbolPairs$Cursor.<init>
[info]   5.3%   892  +     9    Total interpreted (including elided)
[info]
[info]      Compiled + native   Method
[info]   2.4%   410  +     5    scala.reflect.internal.Trees.itransform
[info]   1.8%   304  +     0    scala.reflect.internal.Symbols$Symbol.info
[info]   1.1%   193  +     0    scala.reflect.internal.tpe.TypeMaps$SubstMap.apply
[info]   1.1%   190  +     0    scala.tools.nsc.typechecker.Typers$Typer.runTyper$1
[info]   0.7%   114  +     0    scala.reflect.internal.Symbols$Symbol.typeParams
[info]   0.7%   114  +     0    scala.reflect.internal.Types$TypeRef.equals
[info]   0.6%   108  +     0    scala.reflect.api.Trees$Transformer.transformStats
[info]   0.6%    98  +     0    scala.reflect.internal.Trees$Apply.symbol
[info]   0.6%    97  +     0    scala.reflect.internal.SymbolTable.itraverse
[info]   0.5%    89  +     0    scala.tools.nsc.backend.jvm.analysis.BackendUtils.computeMaxLocalsMaxStack
[info]   0.5%    87  +     0    scala.collection.concurrent.INode.rec_lookup
[info]   0.5%    81  +     0    scala.reflect.internal.tpe.FindMembers$FindMemberBase.walkBaseClasses
[info]   0.4%    71  +     0    scala.tools.asm.tree.analysis.Analyzer.analyze
[info]   0.4%    70  +     0    scala.reflect.internal.Trees$TypeTree.symbol
[info]   0.4%    70  +     0    scala.reflect.internal.transform.UnCurry$$anon$1.apply
[info]   0.4%    66  +     3    scala.reflect.internal.tpe.TypeMaps$TypeMap.mapOver
[info]   0.4%    67  +     0    scala.reflect.internal.tpe.TypeMaps$AsSeenFromMap.apply
[info]   0.4%    65  +     1    scala.reflect.internal.BaseTypeSeqs.compoundBaseTypeSeq
[info]   0.4%    62  +     0    scala.tools.nsc.transform.TypingTransformers$TypingTransformer.atOwner
[info]   0.4%    61  +     1    scala.tools.nsc.typechecker.RefChecks$RefCheckTransformer.transform
[info]   0.3%    58  +     1    scala.reflect.internal.Symbols$Symbol.rawInfo
[info]   0.3%    56  +     2    scala.reflect.internal.Types$TypeRef.normalize
[info]   0.3%    56  +     2    scala.tools.nsc.typechecker.Contexts$Context.lookupSymbol
[info]   0.3%    52  +     1    scala.reflect.internal.transform.Erasure$ErasureMap.apply
[info]   0.3%    53  +     0    scala.reflect.internal.Types$TypeRef$.apply
[info]  54.3%  8908  +   331    Total compiled (including elided)
[info]
[info]          Stub + native   Method
[info]  21.4%     0  +  3634    java.lang.Object.hashCode
[info]  11.1%     0  +  1896    java.lang.System.identityHashCode
[info]   1.6%     0  +   270    java.util.zip.Inflater.inflateBytes
[info]   1.3%     0  +   226    sun.nio.fs.UnixNativeDispatcher.open0
[info]   1.2%     0  +   212    java.util.zip.ZipFile.getEntry
[info]   1.0%     0  +   166    sun.nio.ch.FileDispatcherImpl.close0
[info]   0.6%     4  +    96    java.io.UnixFileSystem.getBooleanAttributes0
[info]   0.4%     0  +    70    java.util.zip.ZipFile.read
[info]   0.3%     0  +    56    java.lang.String.intern
[info]   0.2%     0  +    39    sun.nio.ch.FileDispatcherImpl.write0
[info]   0.2%    32  +     6    java.lang.ClassLoader.defineClass1
[info]   0.2%     0  +    35    java.lang.Throwable.fillInStackTrace
[info]   0.1%     0  +    22    java.lang.Class.getComponentType
[info]   0.1%     0  +    20    sun.misc.Unsafe.defineAnonymousClass
[info]   0.1%     0  +    17    java.lang.Object.clone
[info]   0.1%     0  +    17    java.security.AccessController.doPrivileged
[info]   0.1%     0  +    13    java.lang.reflect.Array.newArray
[info]   0.1%     0  +     9    java.io.UnixFileSystem.list
[info]   0.0%     0  +     6    java.util.zip.Inflater.reset
[info]   0.0%     0  +     4    java.util.zip.ZipFile.freeEntry
[info]   0.0%     0  +     4    java.util.zip.ZipFile.getNextEntry
[info]   0.0%     0  +     3    sun.misc.Unsafe.copyMemory
[info]   0.0%     0  +     3    java.lang.Object.getClass
[info]   0.0%     0  +     3    java.lang.Class.isPrimitive
[info]   0.0%     0  +     3    sun.nio.fs.UnixNativeDispatcher.stat0
[info]  40.4%    36  +  6845    Total stub (including elided)
[info]
[info]   Thread-local ticks:
[info]   0.1%     9             Blocked (of total)
[info]
[info]
[info] Flat profile of 210.05 secs (17045 total ticks): main
[info]
[info]   Interpreted + native   Method
[info]  15.8%     2  +     1    java.lang.ClassLoader.defineClass1
[info]   5.3%     0  +     1    java.lang.Class.getDeclaredMethods0
[info]   5.3%     0  +     1    sun.misc.Unsafe.putInt
[info]   5.3%     0  +     1    sun.misc.Unsafe.defineAnonymousClass
[info]   5.3%     0  +     1    sun.misc.Unsafe.park
[info]   5.3%     0  +     1    sun.misc.Unsafe.defineClass
[info]   5.3%     0  +     1    java.net.SocketInputStream.socketRead0
[info]   5.3%     1  +     0    org.openjdk.jmh.runner.ForkedMain$ShutdownTimeoutThread.<clinit>
[info]   5.3%     1  +     0    java.io.ObjectInputStream$HandleTable.assign
[info]   5.3%     1  +     0    sun.reflect.UTF8.encode
[info]   5.3%     1  +     0    java.lang.invoke.InvokerBytecodeGenerator.emitStaticInvoke
[info]   5.3%     1  +     0    java.util.concurrent.ConcurrentHashMap.transfer
[info]   5.3%     0  +     1    java.util.zip.ZipFile.getEntryTime
[info]  78.9%     7  +     8    Total interpreted
[info]
[info]          Stub + native   Method
[info]  10.5%     0  +     2    java.lang.System.arraycopy
[info]   5.3%     0  +     1    java.lang.Thread.currentThread
[info]  15.8%     0  +     3    Total stub
[info]
[info]   Thread-local ticks:
[info]  99.9% 17026             Blocked (of total)
[info]   5.3%     1             Class loader
[info]
[info]
[info] Global summary of 210.05 seconds:
[info] 100.0% 17194             Received ticks
[info]   0.8%   142             Received GC ticks
[info]  53.7%  9233             Compilation
[info]   0.0%     8             Other VM operations
[info]   0.0%     1             Class loader
```

Comment (apangin) [reply src/symbols_macos.cpp]: `Unknown (native)` typically represents JIT compiler threads, while `Non-Java` usually counts GC and VM threads. If Java code calls a native method (I/O, Zip or whatever), then profiler will correctly show the stack below the last Java frame.

Comment (apangin) [reply src/symbols_macos.cpp]: macOS port does not currently show native frames, but I'm investigating how to implement this. However, Java threads will show up in the profile even if they do a lot of native. Just the top of stack above the first non-Java frame will be cut off.

Comment (retronym) [reply src/symbols_macos.cpp]: Thanks for the clarification. JIT is indeed a major part of our profile (as seen by profiling the same benchmarks on Linux). So the Mac results make sense.

**Pull request merged**